src/topology/sober.lean

```diff
@@ -0,0 +1,282 @@
+def specialization_preorder : preorder α :=
+{ le := λ x y, y ⤳ x,
+  le_refl := λ x, specializes_refl x,
+  le_trans := λ _ _ _ h₁ h₂, specializes.trans h₂ h₁ }
+
+local attribute [instance] specialization_preorder
+
+/-- Specialization forms a partial order on a t0 topological space. -/
+def specialization_order [t0_space α] : partial_order α :=
+{ le_antisymm := λ _ _ h₁ h₂, specializes_antisymm _ _ h₂ h₁,
+  .. specialization_preorder α }
+
+variable {α}
+
+lemma specialization_order.monotone_of_continuous (f : α → β) (hf : continuous f) : monotone f :=
+λ x y h, specializes.map h ⟨f, hf⟩
+
+end specialize_order
+
+section generic_point
+
+/-- `x` is a generic point of `S` if `S` is the closure of `x`. -/
+def is_generic_point (x : α) (S : set α) : Prop := closure ({x} : set α) = S
+
+lemma is_generic_point_def {x : α} {S : set α} : is_generic_point x S ↔ closure ({x} : set α) = S :=
+iff.rfl
+
+lemma is_generic_point.def {x : α} {S : set α} (h : is_generic_point x S) :
+  closure ({x} : set α) = S := h
+
+lemma is_generic_point_closure {x : α} : is_generic_point x (closure ({x} : set α)) := refl _
+
+variables {x : α} {S : set α} (h : is_generic_point x S)
+
+include h
+
+lemma is_generic_point.specializes {y : α} (h' : y ∈ S) :
+  x ⤳ y := by rwa ← h.def at h'
+
+lemma is_generic_point.mem : x ∈ S :=
+h.def ▸ subset_closure (set.mem_singleton x)
+
+lemma is_generic_point.is_closed : is_closed S :=
+h.def ▸ is_closed_closure
+
+lemma is_generic_point.is_irreducible : is_irreducible S :=
+h.def ▸ is_irreducible_singleton.closure
+
+lemma is_generic_point.eq [t0_space α] {y : α} (h' : is_generic_point y S) : x = y :=
+specializes_antisymm _ _ (h.specializes h'.mem) (h'.specializes h.mem)
+
+lemma is_generic_point.mem_open_set_iff
+  {U : set α} (hU : is_open U) : x ∈ U ↔ (S ∩ U).nonempty :=
+⟨λ h', ⟨x, h.mem, h'⟩,
+    λ h', specializes_iff_forall_open.mp (h.specializes h'.some_spec.1) U hU h'.some_spec.2⟩
+
+lemma is_generic_point.disjoint_iff
+  {U : set α} (hU : is_open U) : disjoint S U ↔ x ∉ U :=
+by rw [h.mem_open_set_iff hU, ← set.ne_empty_iff_nonempty, not_not, set.disjoint_iff_inter_eq_empty]
+
+lemma is_generic_point.mem_closed_set_iff
+  {Z : set α} (hZ : is_closed Z) : x ∈ Z ↔ S ⊆ Z :=
+by rw [← is_generic_point_def.mp h, hZ.closure_subset_iff, set.singleton_subset_iff]
+
+lemma is_generic_point.image {f : α → β} (hf : continuous f) :
+  is_generic_point (f x) (closure (f '' S)) :=
+begin
+  rw [is_generic_point_def, ← is_generic_point_def.mp h],
+  apply le_antisymm,
+  { exact closure_mono
+      (set.singleton_subset_iff.mpr ⟨_, subset_closure $ set.mem_singleton x, rfl⟩) },
+  { convert is_closed_closure.closure_subset_iff.mpr (image_closure_subset_closure_image hf),
+    rw set.image_singleton }
+end
+
+end generic_point
+
+section sober
+
+/-- A space is sober if every irreducible closed subset has a generic point. -/
+@[mk_iff]
+class sober_space (α : Type*) [topological_space α] : Prop :=
```

Comment:
As I suggested above, I would call this a `presober_space`. And then a `sober_space` could extend `presober_space` and `t0_space`. (Or we just write both assumptions when we need to.)